File: Subarray Sums Divisible by K.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        def add(i, j):
            nonlocal k
            i += j
            if i < 0:
                i += k
            return i % k

        n = len(nums)

        prefixes = [0] * n
        suffixes = [0] * (n+1)
        total = 0
        for i in range(n):
            total = add(total, nums[i])
            prefixes[i] = total

        t = 0
        for i in range(n-1, -1, -1):
            t = add(t, nums[i])
            suffixes[i] = t

        seen = {0: 1}
        count = 0
        for i in range(1, n+1):
            finding = add(total, -suffixes[i])
            if finding in seen:
                count += seen[finding]
            if prefixes[i-1] not in seen:
                seen[prefixes[i-1]] = 1
            else:
                seen[prefixes[i-1]] += 1
        return count

File: Subarray Sums Divisible by K.cpp
class Solution {
public:
    int add(int i, int j, int& k) {
        i += j;
        while (i < 0) i += k;
        return i % k;
    }

    int subarraysDivByK(vector<int>& nums, int k) {
        int n = nums.size();

        vector<int> prefixes(n, 0);
        vector<int> suffixes(n+1, 0);
        int total = 0;
        for (int i = 0; i < n; i++) {
            total = add(total, nums[i], k);
            prefixes[i] = total;
        }

        int t = 0;
        for (int i = n-1; i >= 0; i--) {
            t = add(t, nums[i], k);
            suffixes[i] = t;
        }

        int count = 0;
        unordered_map<int, int> seen;
        seen[0] = 1;
        for (int i = 1; i <= n; i++) {
            int finding = add(total, -suffixes[i], k);
            count += seen[finding];
            seen[prefixes[i-1]]++;
        }
        return count;
    }
};


